app/org/sagebionetworks/bridge/services/SharedModuleMetadataService.java

```diff
@@ -0,0 +1,208 @@
+package org.sagebionetworks.bridge.services;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import org.sagebionetworks.bridge.dao.SharedModuleMetadataDao;
+import org.sagebionetworks.bridge.exceptions.BadRequestException;
+import org.sagebionetworks.bridge.exceptions.EntityNotFoundException;
+import org.sagebionetworks.bridge.models.sharedmodules.SharedModuleMetadata;
+import org.sagebionetworks.bridge.validators.SharedModuleMetadataValidator;
+import org.sagebionetworks.bridge.validators.Validate;
+
+/** Service for Shared Module Metadata. */
+@Component
+public class SharedModuleMetadataService {
+    private SharedModuleMetadataDao metadataDao;
+
+    /** Shared Module Metadata DAO, configured by Spring. */
+    @Autowired
+    public final void setMetadataDao(SharedModuleMetadataDao metadataDao) {
+        this.metadataDao = metadataDao;
+    }
+
+    /**
+     * Creates the specified metadata object. Gets the module ID from the metadata object. If the module version is not
+     * specified, it will auto-increment one from the most recent version. Throws BadRequestException if the module ID
+     * is not specified. Throws InvalidEntityException if the module metadata is invalid.
+     */
+    public SharedModuleMetadata createMetadata(SharedModuleMetadata metadata) {
+        // These are guaranteed not null by the Controller. These null checks are just to defend against bad coding.s
+        checkNotNull(metadata, "metadata must be non-null");
+
+        // Module ID is validated by getLatestMetadataVersion()
+
+        // Set version if needed. 0 represents an unset module version
+        int oldVersion = getLatestMetadataVersion(metadata.getId());
+        if (metadata.getVersion() == 0) {
+            metadata.setVersion(oldVersion + 1);
+        }
```

Comment:
What if user submits non-zero version that conflicts with an existing id/version pair?